### PR TITLE
Prometheus: Incremental querying option for `to: now` dashboards

### PR DESCRIPTION
**Pull request by leeoniya**
This PR introduces incremental querying to the Prometheus data source. 
[Design doc](https://docs.google.com/document/d/1EQD5PzKozKYVdEBWgwtQFXfCkmpINhqyhTpsOskU3xc/edit#heading=h.xbi0mej5qx7r)
[Spec sheet](https://docs.google.com/document/d/1lGXvZ_XnCFMrN1OKesiau0w83U4x05df0OufmB0m0dk/edit)

Incremental querying is being used to refer to a caching/amending concept that can help unify partial queries for Prometheus and Loki (https://github.com/grafana/grafana/pull/62767) that works for exemplars as well. This hasn't yet been applied to annotations in this PR, but should also work.

In other words, this makes subsequent dashboard queries much smaller, as we're only querying the duration between each refresh, plus any additional overlap that the user has specified (defaults to 10 minutes).

This PR introduces data source configuration fields that default off, so users will need to opt-in data sources to use this feature, which should be considered in beta.

### CODE REVIEW

**Comment by owensmallwood** [comment]
Nice PR 🔥 The query caching team is interested! I've had to do similar things to this in the past, and I found diagrams for all the cases really helped myself and future readers. So I made one! Hopefully I got it right. Let me know 👍 

Screenshot 2023-02-06 at 10 46 17 AM

**Review by matyax — COMMENTED** [review]
Cool! I'm not sure if we need to deal with overlaps (so far it hasn't happened), hopefully not 🤞

**Comment by leeoniya** [comment]
@owensmallwood thanks! :heart_eyes: 

i think it's good to have the result be a combo of red and green to show how they're combined. the last one would be something like 
image

the append/prepend cases result in combined data length, so showing the result as being shorter isnt really accurate.

also i would not call the data "old" but rather "existing". for example, loki will query backwards with most recent data first, so we'd be prepending older data to the newer data. a better nomenclature might be "existing" and "additional".

additionally, this "trims" the combined result to the dashboard time range bounds. so that might be worth showing with vertical dashed lines, but probably too confusing. there can also be time gaps the combined data when there is no overlap, but again might be too confusing to show in a single image.

**Comment by leeoniya** [comment]
> Cool! I'm not sure if we need to deal with overlaps (so far it hasn't happened), hopefully not

@matyax we were told that any data that is newer than 10min may be "unsettled" and can change, so for dashboards like `now-6h` with `refresh: 1min` we actually have to re-request and overwrite at minimum the unsettled `now-10m` of data, even though optimally we'd just need to request the most recent `now-1m`. hopefully this unsettled overlap shrinks in the future as our storage layer provides tighter guarantees.

**Comment by owensmallwood** [comment]
@leeoniya Thanks for clarifying, updated the diagram 👍 Left out trimming. I think it's straightforward to understand without a visualization

**Comment by leeoniya** [comment]
maybe keeping the original colors in the result might work better?

**Comment by leeoniya** [comment]
in the case below, it should be `existing-additional-existing` (additional always overwrites existing in the ranges where they overlap.

image

File: public/app/plugins/datasource/prometheus/QueryCache.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+      // clamp to make sure we don't requery previous 10m when newFrom is ahead of it (e.g. 5min range, 30s refresh)
+      let newFromPartial = Math.max(prevTo! - requeryLastMs, newFrom);
+
+      // console.log(`query previous ${(newTo - newFromPartial) / 1000 / 60} mins`);
+
+      // align the query to the interval to increase the chance of hitting the backend cache
+      // ASK: how to test if this actually has desired effect?
+      // ASK: does the backend already do this anyways? do we need to do it here?
+      // newFromPartial = incrRoundDn(newFromPartial, request.intervalMs);
+      // newToPartial = incrRoundDn(newTo, request.intervalMs);
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/QueryCache.ts:98]: I'll try to help with these questions. I need some more investigation.

Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/QueryCache.ts:98]: thanks!

it's not a huge deal to do it here. but would be good to know :)

Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/QueryCache.ts:98]: @itsmylife had you had a chance to look into these yet? These same questions are relevant to the cacheing PR we're working on as well.

File: public/app/plugins/datasource/prometheus/QueryCache.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+export const getFieldIdent = (field: Field) => `${field.type}|${field.name}|${JSON.stringify(field.labels ?? '')}`;
+
+export class QueryCache {
+  cache = new Map<TargetIdent, TargetCache>();
+
+  // can be used to change full range request to partial, split into multiple requests
+  requestInfo(request: DataQueryRequest<PromQuery>, interpolateString: StringInterpolator): RequestInfo {
+    // TODO: align from/to to interval to increase probability of hitting backend cache
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/QueryCache.ts:37]: Let's open an issue for this

File: packages/grafana-data/src/types/datasource.ts [hunk 3]
```diff
@@ -522,6 +522,10 @@ export interface DataQueryRequest<TQuery extends DataQuery = DataQuery> {
 
   // Make it possible to hide support queries from the inspector
   hideFromInspector?: boolean;
+
+  // For incremental queries that have modified the request, to let the backend know what the requested interval was
+  //@todo, store the delta? Or the stamp?
```
Comment (gtk-grafana) [inline packages/grafana-data/src/types/datasource.ts:527]: I am storing the delta, logic being that this has value without knowing what the range is

File: pkg/services/ngalert/state/cache_test.go [hunk 4]
```diff
@@ -239,8 +239,8 @@ func Test_getOrCreate(t *testing.T) {
 		result := eval.Result{
 			Instance: models.GenerateAlertLabels(5, "result-"),
 			Values: map[string]eval.NumberValueCapture{
-				"A": eval.NumberValueCapture{Var: "A", Value: util.Pointer(1.0)},
-				"B": eval.NumberValueCapture{Var: "B", Value: util.Pointer(2.0)},
+				"A": {Var: "A", Value: util.Pointer(1.0)},
```
Comment (gtk-grafana) [inline pkg/services/ngalert/state/cache_test.go:242]: I didn't change this on purpose, I must have messed something up in merge?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+          ...request.range,
+          from: dateTime(newFromPartial),
+          to: dateTime(newTo),
+        },
+      };
+    } else {
+      reqTargSigs.forEach((targSig, targIdent) => {
+        this.cache.delete(targIdent);
+
+        // TODO: figure out how to purge caches of targets that have permanently stopped making queries
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:109]: This might need to be triggered by the react application, when a dashboard is removed or disabled

File: .betterer.results [hunk 6]
```diff
@@ -5703,6 +5703,11 @@ exports[`better eslint`] = {
       [0, 0, 0, "Unexpected any. Specify a different type.", "1"],
       [0, 0, 0, "Unexpected any. Specify a different type.", "2"]
     ],
+    "public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:5381": [
```
Comment (gtk-grafana) [inline .betterer.results:5706]: I had to assert any[][] as Table, I couldn't find another way around it

File: packages/grafana-data/src/types/datasource.ts (hunk 3 above)
Comment (gtk-grafana) [reply packages/grafana-data/src/types/datasource.ts:527]: We could condense this value further by dividing the delta by the duration, giving us the expected number of samples "saved"

File: packages/grafana-runtime/src/utils/DataSourceWithBackend.ts [hunk 7]
```diff
@@ -204,6 +204,9 @@ class DataSourceWithBackend<
     const headers: Record<string, string> = {};
     headers[PluginRequestHeaders.PluginID] = Array.from(pluginIDs).join(', ');
     headers[PluginRequestHeaders.DatasourceUID] = Array.from(dsUIDs).join(', ');
+    if (request?.originalQueryDelta !== undefined) {
+      headers['X-Grafana-Modified-From'] = request.originalQueryDelta.toString(10);
```
Comment (gtk-grafana) [inline packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:208]: The hope is that we can use this in Faro or in Grafana logs to generate a dashboard displaying how many records we've avoided querying

Comment (leeoniya) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:208]: thinking more about this, let's try a different approach that will hopefully be more easily consumed by leveraging Faro (we can probably do this in a follow-up to keep this PR manageable)

basically, we should try to measure the amount of data and latency during the initial full-range request. we can do this by timing `fetch()` and inspecting the `Content-Length` and `Content-Encoding` headers of the response.

we can then assume that every full-range refresh would have been the same size and the same latency, it may not always be true, but i think it is in most cases. but when we do an incremental request instead, we can take the difference and calculate the "savings" in latency and in transfer size, and send this savings info to Faro. it won;t be exact, but i think it will be close enough.

Comment (leeoniya) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts]: pushed https://github.com/grafana/grafana/commit/0090d6f11b673c0d45e0cf26343c211f225f749b but it looks like the backend does not currently send `Content-Length` headers, so we don't get anything here:

https://github.com/grafana/grafana/commit/0090d6f11b673c0d45e0cf26343c211f225f749b#diff-a38b9903e300e119fb4b391c6bd7332390d562c5c2215e1e6a68afe504e8117bR11

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+type TargetIdent = string;
+
+// query + template variables + interval + raw time range
+// used for full target cache busting -> full range re-query
+type TargetSig = string;
+
+type TimestampMs = number;
+
+type StringInterpolator = (expr: string) => string;
+export const defaultPrometheusQueryOverlapWindowSeconds = 10 * 60 * 1000;
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:17]: this is not seconds. take the 1000 out and multiply it inside.

File: public/app/plugins/datasource/prometheus/types.ts [hunk 9]
```diff
@@ -32,6 +32,8 @@ export interface PromOptions extends DataSourceJsonData {
   prometheusType?: PromApplication;
   prometheusVersion?: string;
   defaultEditor?: QueryEditorMode;
+  incrementalQuerying?: boolean;
+  incrementalQueryOverlapDuration?: number; // in seconds
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/types.ts:36]: why do we call it `duration` in some places and `window` in others? (i prefer `window`, personally)

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+/**
+ * NOMENCLATURE
+ * Target: The request target (DataQueryRequest), i.e. a specific query reference within a panel
+ * Ident: Identity: the string that is not expected to change
+ * Sig: Signature: the string that is expected to change, upon which we wipe the cache fields
+ */
+export class QueryCache {
+  private overlapWindowInSeconds: number;
+  constructor(overlapInSeconds?: number) {
+    this.overlapWindowInSeconds = overlapInSeconds ?? defaultPrometheusQueryOverlapWindowSeconds;
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:57]: `this.queryOverlapWindow`? (also, just match the constructor arg name).

then multiply it here by 1000 to get millis back

File: public/app/plugins/datasource/prometheus/types.ts (hunk 9 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/types.ts:36]: I screwed up the refactor 🤦

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 10 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts]: I just left this in ms

**Comment by gtk-grafana** [comment]
Todo: Remove ms from FE

File: .betterer.results (hunk 6 above)
Comment (joshhunt) [reply .betterer.results:5462]: The goal of betterer is to remove the checks once we've resolved the issues, but we can't do that if we keep adding new exceptions to it.

If we are comitting to having unsafe code, you must "pay the cost" of that inline, using the `// eslint-ignore-next-line` comments, rather than adding new debt here to resolve.

Comment (gtk-grafana) [reply .betterer.results:5462]: `// eslint-ignore-next-line` doesn't appear to stop the betterer results from getting added in here when I run update?

Comment (gtk-grafana) [reply .betterer.results:5462]: I committed the changes after adding `// eslint-ignore-next-line` to all the type assertions and running betterer -u, still looks like it's polluting this file. Do you know what I'm doing wrong?

**Comment by leeoniya** [comment]
> This hasn't yet been applied to annotations in this PR, but should also work.

is the plan to do this in a follow-up or part of this PR?

**Comment by gtk-grafana** [comment]
@leeoniya I was planning in a follow up, I added a placeholder to the epic

File: .betterer.results (hunk 6 above)
Comment (joshhunt) [reply .betterer.results:5464]: Sorry, correct comment is

```ts
// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
const hello = {} as FolderInfo;
```

image

File: docs/sources/administration/provisioning/index.md [hunk 11]
```diff
@@ -169,56 +169,58 @@ Common settings in the [built-in core data sources]({{< relref "../../datasource
...
-| prometheusType             | string  | Prometheus                                                       | The type of the Prometheus data sources. such as `Prometheus`, `Cortex`, `Thanos`, `Mimir`                                                        |
-| implementation             | string  | AlertManager                                                     | The implementation of the AlertManager data source, such as `prometheus`, `cortex` or `mimir`                                                     |
-| handleGrafanaManagedAlerts | boolean | AlertManager                                                     | When enabled, Grafana-managed alerts are sent to this Alertmanager                                                                                |
+| Name                          | Type    | Data source                                                      | Description                                                                                                                                           |
```
Comment (gtk-grafana) [inline docs/sources/administration/provisioning/index.md:172]: Prettier forces me to update a bunch of whitespace, I apologize for all of the conflicts that this will inevitably introduce

File: public/app/features/live/data/amendTimeSeries.ts
Comment (itsmylife) [inline public/app/features/live/data/amendTimeSeries.ts:1]: I think having unit tests for these functions might be useful for the future.

Comment (itsmylife) [reply public/app/features/live/data/amendTimeSeries.ts:1]: Also could you please clear the `console.log`s and unnecessary comment lines?

File: public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx [hunk 12]
```diff
@@ -335,6 +337,50 @@ export const PromSettings = (props: Props) => {
             />
           </div>
         </div>
+
+        <div className="gf-form-inline">
+          <div className="gf-form max-width-30">
+            <FormField
+              label="Incremental querying (beta)"
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx:344]: Just asking, should we use [this](https://developers.grafana.com/ui/latest/index.html?path=/story/data-display-badge--basic) to show beta notice instead of having `(beta)`?

File: public/app/plugins/datasource/prometheus/datasource.tsx [hunk 13]
```diff
@@ -79,6 +80,7 @@ export class PrometheusDatasource
   type: string;
   editorSrc: string;
   ruleMappings: { [index: string]: string };
+  hasIncrementalQuery?: boolean;
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/datasource.tsx:83]: Do we need this as optional?

File: public/app/plugins/datasource/prometheus/datasource.tsx [hunk 14]
```diff
@@ -445,14 +452,30 @@ export class PrometheusDatasource
...
+            ...response,
+            data: this.cache.procFrames(request, requestInfo, response.data),
+          };
+          return transformV2(amendedResponse, request, {
+            exemplarTraceIdDestinations: this.exemplarTraceIdDestinations,
+          });
+        }),
         tap((response: DataQueryResponse) => {
           trackQuery(response, request, startTime);
+          //@todo track incremental query savings
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/datasource.tsx:478]: Is [this](https://github.com/grafana/grafana/issues/65214) tracking this todo?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.test.ts
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.test.ts:1]: 🚀

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+      const duration = parseDuration(defaultPrometheusQueryOverlapWindow);
+      this.overlapWindow = durationToMilliseconds(duration);
+    }
+  }
+
+  cache = new Map<TargetIdent, TargetCache>();
+
+  // can be used to change full range request to partial, split into multiple requests
+  requestInfo(request: DataQueryRequest<PromQuery>, interpolateString: StringInterpolator): CacheRequestInfo {
+    // TODO: align from/to to interval to increase probability of hitting backend cache
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:83]: Are we tracking this `todo` somewhere? If not could you please create an issue?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+        doPartialQuery = newTo > prevTo && newFrom <= prevTo;
+      }
+
+      if (!doPartialQuery) {
+        break;
+      }
+    }
+
+    // console.log(`${doPartialQuery ? 'partial' : 'full'} query`);
+
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:125]:
```suggestion

```

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+        let cachedFrames = (targIdent ? this.cache.get(targIdent)?.frames : null) ?? [];
+
+        respFrames.forEach((respFrame: DataFrame) => {
+          // skip empty frames
+          if (respFrame.length === 0 || respFrame.fields.length === 0) {
+            return;
+          }
+
+          // frames are identified by their second (non-time) field's name + labels
+          // TODO: maybe also frame.meta.type?
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:192]: This should also be trackable in an issue. Could you please create one?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+        this.cache.set(targIdent, {
+          sig: requestInfo.targSigs.get(targIdent)!,
+          frames: nonEmptyCachedFrames,
+          prevTo: newTo,
+        });
+
+        outFrames.push(...nonEmptyCachedFrames);
+      });
+
+      // transformV2 mutates field values for heatmap de-accum, and modifies field order, so we gotta clone here, for now :(
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:245]: Should we track this in an issue to make it better?

File: public/app/plugins/datasource/prometheus/types.ts [hunk 19]
```diff
@@ -32,6 +32,8 @@ export interface PromOptions extends DataSourceJsonData {
   prometheusType?: PromApplication;
   prometheusVersion?: string;
   defaultEditor?: QueryEditorMode;
+  incrementalQuerying?: boolean;
+  incrementalQueryOverlapWindow?: string; // ms
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/types.ts:36]: I don't like to comment line but making that information explicit in the naming is not likable either.

**Review by itsmylife — COMMENTED** [review]
Just put lots of small questions and suggestions.

**Comment by itsmylife** [comment]
I tried this:
- I set the time frame now-2d (Last 2 days)
- I picked a metric and run it. Got 60kB data
- After a few seconds I run it again and got 1.9kB
- I picked another metric and run it. Got 20kB data
- After a few seconds I run it again and got 935B data
- I run the first metric again. Got 60kB data.

My question is shouldn't I get 1.9kB again or at least less than 60kB?

File: public/app/plugins/datasource/prometheus/types.ts (hunk 19 above)
Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/types.ts:36]: this comment is no longer true, actually. so it can be removed. we now parse these time strings, so it can be whatever.

**Comment by leeoniya** [comment]
> My question is shouldn't I get 1.9kB again or at least less than 60kB?

this is by design. we invalidate the cache whenever the same panel sends a different query, otherwise we have much bigger memory leak risk and potentially hold too much in memory that will need to be continuously monitored and evicted. we're planning for this cache to primarily by for non-explore type cases, where you just watch an auto-refreshing dashboard where queries stay stable.

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 18 above)
Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:245]: it might be expensive to make copies continuously, so i'm not sure this is something that can/should be "fixed". but maybe.

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 17 above)
Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:192]: i'm not sure if this is a real todo, it was just a possible thought in case we didnt have enough determinism here. if everything works okay then we can just remove it, but would also be useful to leave the comment in for a while so it's easy to find exact place to adjust.

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 15 above)
Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:83]: i think we are waiting on your assessment if it has any effect ;)

Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:83]: I added an item to the epic to address this, I think we can come back to this

File: public/app/plugins/datasource/prometheus/types.ts (hunk 19 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/types.ts:36]: Yeah this should be removed, and has been!

File: public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx (hunk 12 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx:344]: image
I would love to! But where would we put it? We can't add JSX to the input labels without refactoring a bunch of stuff. I added it after the field, but I dunno if that looks any better. @catherineymgui?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 15 above)
Comment (itsmylife) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:83]: @leeoniya Ah yes I remember. I forgot that one sorry. As @gtk-grafana suggested let's have it in the epic and come back to it later on

**Comment by itsmylife** [comment]
@gtk-grafana `requestId` was removed intentionally. See https://github.com/grafana/grafana/pull/63913/files#diff-df1f62e68ea660d4a9d411e360d9d3cc554bd877b990ebd8684404d01c0504da

**Review by itsmylife — APPROVED** [review]
🚀 🚀 🚀 
Thanks for the amazing work! @gtk-grafana @leeoniya

File: docs/sources/datasources/prometheus/_index.md [hunk 20]
```diff
@@ -172,3 +174,11 @@ Grafana lists these variables in dropdown select boxes at the top of the dashboa
 Grafana refers to such variables as template variables.
 
 For details, see the [template variables documentation]({{< relref "./template-variables/" >}}).
+
+## Incremental Dashboard Queries (beta)
+
+As of Grafana 9.5, data sources can be configured to query dashboards incrementally, instead of querying the entire duration on each dashboard refresh.
```
Comment (leeoniya) [inline docs/sources/datasources/prometheus/_index.md:180]: Grafana 10, right?

File: docs/sources/datasources/prometheus/_index.md [hunk 21]
```diff
@@ -175,3 +177,11 @@ Grafana lists these variables in dropdown select boxes at the top of the dashboa
 Grafana refers to such variables as template variables.
 
 For details, see the [template variables documentation]({{< relref "./template-variables/" >}}).
+
+## Incremental Dashboard Queries (beta)
+
+As of Grafana 10, data sources can be configured to query live dashboards incrementally, instead of re-querying the entire duration on each dashboard refresh.
```
Comment (gtk-grafana) [inline docs/sources/datasources/prometheus/_index.md:183]:
```suggestion
As of Grafana 10, the Prometheus data source can be configured to query live dashboards incrementally, instead of re-querying the entire duration on each dashboard refresh.
```

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+    // all targets are queried together, so we check for any that causes group cache invalidation & full re-query
+    let doPartialQuery = shouldCache;
+    let prevTo: TimestampMs;
+
+    // pre-compute reqTargSigs
+    const reqTargSigs = new Map<TargetIdent, TargetSig>();
+    request.targets.forEach((targ) => {
+      let targIdent = `${request.dashboardUID}|${request.panelId}|${targ.refId}`;
+      let targExpr = interpolateString(targ.expr);
+      let targSig = getTargSig(targExpr, request, targ); // ${request.maxDataPoints} ?
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:99]: https://github.com/grafana/grafana/pull/65952/files#diff-79a8e0c726aee282de46b93b30ea4022261351ea9c0d46e6f988296cfcf5c900R173-R179

**Review by gtk-grafana — APPROVED** [review]
Looks good, great work everybody!

**Comment by nemobis** [comment]
> [Design doc](https://docs.google.com/document/d/1EQD5PzKozKYVdEBWgwtQFXfCkmpINhqyhTpsOskU3xc/edit#heading=h.xbi0mej5qx7r)
> [Spec sheet](https://docs.google.com/document/d/1lGXvZ_XnCFMrN1OKesiau0w83U4x05df0OufmB0m0dk/edit)

Is there a public version of these documents? Ideally, please attach an ODT copy here. Thanks! :)
